src/NuGet.Clients/VsConsole/Console/ScriptExecutor.cs

```diff
@@ -0,0 +1,328 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.ComponentModel.Composition;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Configuration;
+using NuGet.Frameworks;
+using NuGet.PackageManagement;
+using NuGet.PackageManagement.PowerShellCmdlets;
+using NuGet.PackageManagement.VisualStudio;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+using EnvDTEProject = EnvDTE.Project;
+using Strings = NuGet.ProjectManagement.Strings;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGetConsole
+{
+    [Export(typeof(IScriptExecutor))]
+    public class ScriptExecutor : IScriptExecutor
+    {
+        private AsyncLazy<IHost> Host { get; }
+        private ISolutionManager SolutionManager { get; }  = ServiceLocator.GetInstance<ISolutionManager>();
+        private string SolutionDirectory { get; set; }
+        private ISettings Settings { get; } = ServiceLocator.GetInstance<ISettings>();
+        private ConcurrentDictionary<PackageIdentity, bool> InitScriptExecutions
+            = new ConcurrentDictionary<PackageIdentity, bool>(PackageIdentityComparer.Default);
+
+        public ScriptExecutor()
+        {
+            Host = new AsyncLazy<IHost>(GetHostAsync, ThreadHelper.JoinableTaskFactory);
+            Reset();
+        }
+
+        [Import]
+        public IPowerConsoleWindow PowerConsoleWindow { get; set; }
+
+        [Import]
+        public IOutputConsoleProvider OutputConsoleProvider { get; set; }
+
+        public void Reset()
```

Comment:
Lets talk about this, reset does something very specific, it looks like it keeps a list of only init.ps1 that didn't run. It seems odd, and I don't quite understand what it is for. So I'd like to understand the algorithm, the purpose etc. in comments or see if it needs to stay at all in its current form.